lib/stm32/common/spi_common_v3.c

```diff
@@ -0,0 +1,834 @@
+@returns data Unsigned int16. 8 or 16 bit data.
+*/
+
+uint16_t spi_xfer(uint32_t spi, uint16_t data)
+{
+	spi_write(spi, data);
+
+	/* Wait for transfer finished. */
+	while (!(SPI_SR(spi) & SPI_SR_RXP));
+
+	/* Read the data (8 or 16 bits, depending on DFF bit) from DR. */
+	return SPI_RXDR(spi);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Set Bidirectional Simplex Receive Only Mode.
+
+The SPI peripheral is set for bidirectional transfers in two-wire simplex mode
+(using a clock wire and a bidirectional data wire), and is placed in a receive
+state.
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_set_bidirectional_receive_only_mode(uint32_t spi)
+{
+	SPI_CFG2(spi) = (SPI_CFG2(spi) & ~(SPI_CFG2_COMM_MASK << SPI_CFG2_COMM_SHIFT))
+		| (SPI_CFG2_COMM_HALF_DUPLEX << SPI_CFG2_COMM_SHIFT);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Set Bidirectional Simplex Receive Only Mode.
+
+The SPI peripheral is set for bidirectional transfers in two-wire simplex mode
+(using a clock wire and a bidirectional data wire), and is placed in a transmit
+state.
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_set_bidirectional_transmit_only_mode(uint32_t spi)
+{
+	SPI_CFG2(spi) = (SPI_CFG2(spi) & ~(SPI_CFG2_COMM_MASK << SPI_CFG2_COMM_SHIFT))
+		| (SPI_CFG2_COMM_HALF_DUPLEX << SPI_CFG2_COMM_SHIFT);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Enable the CRC.
+
+The SPI peripheral is set to use a CRC field for transmit and receive.
+Note that in order to use CRC, you have to configure transfer size before
+data transfer actually starts on SPI bus. See @ref spi_set_transfer_size().
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_enable_crc(uint32_t spi)
+{
+	SPI_CFG1(spi) |= SPI_CFG1_CRCEN;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Disable the CRC.
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_disable_crc(uint32_t spi)
+{
+	SPI_CFG1(spi) &= ~SPI_CFG1_CRCEN;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI set length of next transfer
+ *
+ * Sets frame size of next transmission. It may be either transmission or
+ * reception. If enabled, then length includes CRC. It is mandatory to set
+ * transfer length before CRC is enabled and data transfer is started.
+ *
+ * @param[in] spi SPI peripheral identifier @ref spi_reg_base.
+ * @param[in] size size of next frame. If given as 0xFFFF, then transfer is
+ * infinite. Cant be used with CRC enabled.
+ */
+void spi_set_transfer_size(uint32_t spi, uint16_t size)
+{
+	/* This code works and I don't have a clue why */
```

Comment:
well, that's confidence inspiring :)

Reply:
"EOT is set by hardware as soon as a full transfer is complete"
"In master, EOT event terminates the data transaction and handles SS output optionally. In both master and slave, the event handles CRC transaction."
"EOT is cleared by software write 1 to EOTC bit at SPI2S_IFCR."

So if the previous transfer has completed, this bit is clear, probably to allow starting a new transfer? Not sure why this is required, this bit should be cleared before a transaction...